### PR TITLE
fixed issue #546 - broken multi column sort with invisible columns

### PR DESCRIPTION
**Pull request by arashdalir**
added new property `allColumns` and 3 new functions `setAllColumns(columnDefinitions)` `getAllColumns(defaultToColumns)` and getColumnById(columnId, useVisibleColumns)` - the last function is used in `setupColumnSort` in order to get correct column: if `allColumns ` is set then use `allColumns` else use `columns` as reference list

### CODE REVIEW

**Comment by arashdalir** [comment]
@ghiscoding , @6pac this is a basic solution based on our discussion on #546! I have a fundamental issue with this solution though! if a user updates `columns` by using `setColumns`, the change is not necessarily reflected in `allColumns` which will mean that using my `getColumnById` might not return the correct `column`. I believe a more elegant solution is to somehow keep `columns` and `allColumns` synchronised in case properties of one changes. 
I still believe the only "right" solution is to add a property named `visible: true` to the columns and adapt the functions needed for rendering etc. to consider that attribute. otherwise the values in `allColumns` and `columns` with drift apart fast if `setColumns` or `setAllColumns` is set often...

**Comment by arashdalir** [comment]
or well, another solution is to add a `setVisibleColumns(columnIds)` function to `slick.grid` which will internally manage the state of `allColumns` and `columns`. in that case, using `setColumns` actually sets both `allColumns` and `columns` values, and `setVisibleColumns` reduces the set whenever called. in case a `setColumns` is called after `setVisibleColumns`, they are synchronized again and the user must set the visibility again...

File: slick.grid.js [hunk 1]
```diff
@@ -2765,6 +2784,31 @@ if (typeof Slick === "undefined") {
       }
     }
 
+    function setAllColumns(columnDefinitions){
+      var _columns = columns.slice(0);
```
Comment (ghiscoding) [inline slick.grid.js:2788]: if you just want to make a copy, you can omit the 0 and just go with `.slice()`

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (ghiscoding) [inline composer.json:1]: could you please remove this file from the project, I think it's unnecessary

File: slick.grid.js [hunk 3]
```diff
@@ -2726,6 +2728,23 @@ if (typeof Slick === "undefined") {
       return columns;
     }
 
+    function getColumnById(columnId, useOnlyVisibleColumns){
+      var columnList = useOnlyVisibleColumns?getColumns():getAllColumns(true);
```
Comment (ghiscoding) [inline slick.grid.js:2732]: can you space this code, it would be a bit more readable

**Comment by ghiscoding** [comment]
I would say that in any case we want to avoid pushing breaking changes as much as possible because there are thousands of people using the lib and few hundreds using my libs but I can update mine quickly. However what could potentially be done is to make the `setColumns` intelligent and detect, by the columns length and their position, if the lib should call the old code or the new code of `setVisibleColumns` that you're proposing.  What I mean is that we could propose/add some new functions and if the user doesn't use it then it would use the old code, in some areas we also added some `console.warn` to tell the user this function is deprecated. So there are couple of alternatives you can look at, just try to avoid breaking changes

**Comment by 6pac** [comment]
I've been staying out of this one, mainly because I use my own customised version of the grid, and a different customised DataView that doesn't have a lot of the limitations of the public one.  Not sure if I even have this issue. I'd like to merge but the two have diverged so far that it would be a *lot* of work.

So I'd make two comments:  

1) the easiest workaround would simply be to rewrite the sort code so it skips nonexistent columns

2) ultimately, the underlying data is the important thing. The column are just an overlay. So if the sort code was written to depend on the data properties rather than the columns, we would avoid all this mess. Perhaps it needs to cache the column property name first time around?

File: composer.json (hunk 2 above)
Comment (arashdalir) [reply composer.json:1]: damn! don't understand how this file always gets pushed! it doesn't exist in my code! :/ sorry! will be removed

File: slick.grid.js (hunk 3 above)
Comment (arashdalir) [reply slick.grid.js:2732]: sure

**Comment by arashdalir** [comment]
> I would say that in any case we want to avoid pushing breaking changes as much as possible because there are thousands of people using the lib and few hundreds using my libs but I can update mine quickly. However what could potentially be done is to make the `setColumns` intelligent and detect, by the columns length and their position, if the lib should call the old code or the new code of `setVisibleColumns` that you're proposing. What I mean is that we could propose/add some new functions and if the user doesn't use it then it would use the old code, in some areas we also added some `console.warn` to tell the user this function is deprecated. So there are couple of alternatives you can look at, just try to avoid breaking changes

I was thinking to leave the solution to skip the not found columns in place and instead of adding the `setAllColumns`, and use the `setVisibleColumnsById` as a new feature! for that, the following changes in `grid` would be necessary:
1. add a new `setVisibleColumnsById()` and `getVisibleColumns()`. the `getAllColumns` and `setAllColumns` will be removed.
1. keep the `allColumns` and `columns` in the background and manage their values using `setVisibleColumnsById`
1. change the signature of `getColumns()` to `getColumns(returnAllColumns:boolean=false)`; which will behave the same without the optional parameter (or with false) - i.e. returning visible columns; and return all columns if the parameter is set to true.

if someone doesn't use the new feature, the old logic would still work by changing the complete column set; but the new way would let them have a list of all columns in the background (and get it using `getColumns(true)` for returning visible and invisible columns) or just the visible ones using `getColumns()`. 

we could then extend the `columnPicker` to:
1. also have a `setVisibleColumns` which would call the `setVisibleColumns` in the grid.
1. use `getColumns()` in `getVisibleColumns()` and `getColumns(true)` in `getAllColumns()`.
1. use `setVisibleColumns` function instead of `grid.setColumns()` [here](https://github.com/6pac/SlickGrid/blob/39707b830b270ced2a26436a50092d5d81ebb615/controls/slick.columnpicker.js#L218),

that way, the current functionality will stay unchanged, as the `grid.getColumns()` function will return the visible columns as before and expert users can use `grid.getColumns(true)` to use the new features.

**Comment by arashdalir** [comment]
> I've been staying out of this one, mainly because I use my own customised version of the grid, and a different customised DataView that doesn't have a lot of the limitations of the public one. Not sure if I even have this issue. I'd like to merge but the two have diverged so far that it would be a _lot_ of work.
> 
> So I'd make two comments:
> 
> 1. the easiest workaround would simply be to rewrite the sort code so it skips nonexistent columns
> 2. ultimately, the underlying data is the important thing. The column are just an overlay. So if the sort code was written to depend on the data properties rather than the columns, we would avoid all this mess. Perhaps it needs to cache the column property name first time around?

I'd really like to see your version! :D 

tbh, we are considering changing from `slick-grid` to another solutions like `elastic-ui` or `quasar` because `slick-grid` has some underlying issues which cannot be resolved that easily; like its use of absolute positioning for rows or the way `detailview` handles detail rows or... if your version has resolved these issues, we might be able to create a new `v3` based on it and keep the 2 versions separate: the users of the current version will be able to use their branch and those who will be starting, could use the newer one. just saying...

File: slick.grid.js (hunk 1 above)
Comment (arashdalir) [reply slick.grid.js:2788]: sure, will change - waiting for our conversation to reach an agreement

**Comment by 6pac** [comment]
Unfortunately, those are not things I changed. If anything, the UI capabilities are much less than the current grid. I don't have the split grid or the details view at all in my version (it forked about four years ago before all that).  
What it does have is:
- FlexTable and FlexView, which are an improved DataView, with full undo/redo, change tracking for client/server edit saves, id not needed for row logic (row index is used instead), support for array or property based rows, plus everything DataView does
- a data provider to allow the grid to be interfaced with a plain js object, DataView or FlexTable/FlexView, and any other object that offers the standard interface

**Comment by 6pac** [comment]
Hey folks, is this still relevant or should I close it?

**Comment by arashdalir** [comment]
> Hey folks, is this still relevant or should I close it?

I have actually implemented the solution as described here and it works in my version. but because of this, my deployment branch is not the active main branch anymore.

honestly, I have not checked the newest version to see the issue has been already resolved or not. I suppose, if none of us has worked on it, this will still be an open issue?

**Comment by ghiscoding** [comment]
closing in favor of #723 which is up to date with latest master branch

**Pull request closed without merging**